python/tskit/metadata.py

```diff
@@ -111,7 +111,10 @@ def __init__(self, schema: Mapping[str, Any]) -> None:
         pass
 
     def encode(self, obj: Any) -> bytes:
-        return json.dumps(obj).encode()
+        # Here we sort and indent to give a canonical string that matches
+        # https://github.com/nlohmann/json which is used by C++ programs
+        # that generate tree sequences
+        return json.dumps(obj, sort_keys=True, indent=4).encode()
```

Comment:
Agreed - it looks like the indent is controllable anyway: https://nlohmann.github.io/json/api/basic_json/dump/
I've made the change and removed `indent`.

Reply:
I think we should go a step further though, and remove whitespace:
```
>>> json.dumps({"x": 1, "b": 2})
'{"x": 1, "b": 2}'
>>> json.dumps({"x": 1, "b": 2}, sort_keys=True, separators=(',', ':'))
'{"b":2,"x":1}'
```
and we make the second form the canonicalised JSON. We should probably just have a function ``tskit.canonical_json(d): return json.dumps({"x": 1, "b": 2}, sort_keys=True, separators=(',', ':'))`` so that this is documented and easy to reuse.

Reply:
Yep, good idea. Done.

Reply:
Looks good, thanks. Can we get a few simple unit tests for ``canonical_json`` so we don't regress on this please?

Reply:
Done!

Reply:
Perfect, thanks!